### PR TITLE
Add Deep Learning Blog 2

### PR DESCRIPTION
**Pull request by BenWilson2**

### CODE REVIEW

**Comment by AbeOmor** [comment]
LGTM!

File: website/blog/2024-04-26-deep-learning-part-2/index.md [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+Addressing unfair clauses in Terms of Service (TOS) agreements through machine learning (ML) is particularly relevant due to the pressing
+need for transparency and fairness in legal agreements that affect consumers. Consider the following clause from an example TOS
+agreement: **"We may revise these Terms from time to time. The changes will not be retroactive, and the most current version of the Terms, which will always..."**
+This clause stipulates that the service provider may suspend or terminate the service at any time for any reason,
+with or without notice. Most people would consider this to be quite unfair.
+
+While this sentence is buried quite deeply within a fairly dense document, an ML algorithm is not burdened by the exhaustion that a human
+would have for combing through the text and identifying clauses that might seem a bit unfair. By automatically identifying potentially
+unfair clauses, transformers can help protect consumers from exploitative practices, ensure greater compliance with legal standards,
```
Comment (B-Step62) [inline website/blog/2024-04-26-deep-learning-part-2/index.md:41]: nit: `transformers can help protect customers...` sounds a bit weird protect consumers: sounds a bit weird as transformers us a framework not a model, like we don't say `python can help....`. Can we rephrase it to "the ML model can help ..."?

File: website/blog/2024-04-26-deep-learning-part-2/index.md [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+and foster trust between service providers and users.
+
+A base transformer model, without specialized fine-tuning, faces several challenges in accurately identifying unfair Terms of Service clauses.
```
Comment (B-Step62) [inline website/blog/2024-04-26-deep-learning-part-2/index.md:44]:
```suggestion
A base pretrained model, without specialized fine-tuning, faces several challenges in accurately identifying unfair Terms of Service clauses.
```
nit: transformer is an architecture name

File: website/blog/2024-04-26-deep-learning-part-2/index.md [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+increase in accuracy or effectiveness.
+
+In this context, the use of the **RoBERTa-base** model is particularly effective, provided that it is fine-tuned. This model is robust
+enough to handle complex tasks like discerning embedded instructions within texts, yet it is sufficiently compact to be fine-tuned
+on modest hardware, such as an Nvidia T4 GPU.
+
+### What is PEFT?
+
+[Parameter-Efficient Fine-Tuning (PEFT)](https://huggingface.co/docs/peft/en/index) approaches are advantageous as they involve
+keeping the bulk of the pre-trained model parameters fixed while only training a few additional layers. This methodology not only
```
Comment (B-Step62) [inline website/blog/2024-04-26-deep-learning-part-2/index.md:62]:
```suggestion
keeping the bulk of the pre-trained model parameters fixed while only training a fairy small percentage of parameters.
```
nit: Not all PEFT methods uses additional layers.

File: website/blog/2024-04-26-deep-learning-part-2/index.md [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+            torch.Tensor: The computed loss from the model as a tensor.
+        """
+        return self.model(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            labels=label
+        )
+```
+
+Additional references for the full implementation can be [seen within the companion repository here](https://github.com/puneet-jain159/deeplearning_with_mlfow/blob/master/custom_module/fine_tune_clsify_head.py)
```
Comment (B-Step62) [inline website/blog/2024-04-26-deep-learning-part-2/index.md:143]: nit: not related to the blog content, but the linked code has a bit incorrect Python syntax
https://github.com/puneet-jain159/deeplearning_with_mlfow/blob/master/custom_module/fine_tune_clsify_head.py#L60-L100

There are two `return` in the `predict` function where anything below L77 is simply discarded.

File: website/blog/2024-04-26-deep-learning-part-2/index.md [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+       )
+   ],
+   default_root_dir=train_config.model_checkpoint_dir,
+   fast_dev_run=bool(train_config.debug_mode_sample),
+   max_epochs=train_config.max_epochs,
+   max_time=train_config.max_time,
+   precision="32-true"
+)
+
+trainer.fit(model=nlp_model, datamodule=datamodule)
```
Comment (B-Step62) [inline website/blog/2024-04-26-deep-learning-part-2/index.md:215]: This `nlp_model` is not defined above and suddenly appears. I think it's crucial part as it contains pretrained RoBERTa model name and LoRA configurations. Can we add that code snippet?

File: website/blog/2024-04-26-deep-learning-part-2/index.md [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+
+
+mlflow.pytorch.autolog(disable = True)
+
+run_id = '<Add the run ID>'
+model = mlflow.pytorch.load_model(f"runs:/{run_id}/model")
+
+model = mlflow.pytorch.load_checkpoint(TransformerModule, run_id, 3)
+
+train_module.predict(model, dataloaders = DataLoader(ds))
```
Comment (B-Step62) [inline website/blog/2024-04-26-deep-learning-part-2/index.md:258]: Do we need to do this instead of `model.predict()?`

File: website/blog/2024-04-26-deep-learning-part-2/index.md [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+enhances the reproducibility and efficiency of model development, but also fosters a collaborative environment where insights
+and improvements can be easily shared and implemented.
+
+As we continue to push the boundaries of what these models can achieve, tools like MLflow will be instrumental in harnessing their full potential.
+
+If you're interested in seeing the full example in its entirety, feel free to [see the full example implementation](https://github.com/puneet-jain159/deeplearning_with_mlfow)
+
+### Check out the code
+
+The code we provide will delve into additional aspects such as training from a checkpoint, integrating MLflow and TensorBoard, and utilizing Pyfunc for model wrapping, among others. These resources are specifically tailored for implementation on [Databricks Community Edition](https://mlflow.org/blog/databricks-ce).
```
Comment (B-Step62) [inline website/blog/2024-04-26-deep-learning-part-2/index.md:275]: nit: Can we add link to the repo here? (Ideally a notebook that users can import to DCE)

**Review by B-Step62 — CHANGES_REQUESTED** [review]
Overall looks good, but can we address [this comment](https://github.com/mlflow/mlflow-website/pull/61#discussion_r1584052383) before merging?

File: website/blog/2024-04-26-deep-learning-part-2/index.md (hunk 4 above)
Comment (puneet-jain159) [reply website/blog/2024-04-26-deep-learning-part-2/index.md:143]: Hi @B-Step62  
Corrected the code.  and tested the updated code with an end 2 end run. 

Link : https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/4093282202515276

File: website/blog/2024-04-26-deep-learning-part-2/index.md (hunk 6 above)
Comment (puneet-jain159) [reply website/blog/2024-04-26-deep-learning-part-2/index.md:258]: Yes. because the loaded object is a Pytorch object (not a pyfunc object) which does not have a predict method. 

`type(model)`
> custom_module.fine_tune_clsify_head.TransformerModule

**Review by B-Step62 — APPROVED** [review]
LGTM! Thank you for the udpate!!
